compiler/rustc_mir_build/src/check_unsafety.rs

```diff
@@ -599,13 +599,10 @@ pub fn check_unsafety<'tcx>(tcx: TyCtxt<'tcx>, def: ty::WithOptConstParam<LocalD
 
     // Closures are handled by their owner, if it has a body
     if tcx.is_closure(def.did.to_def_id()) {
-        let owner = tcx.hir().local_def_id_to_hir_id(def.did).owner;
```

Comment:
just to be clear I understand, because I understood something different from your text. `owner` here is `== def.did`, and thus invoking `thir_check_unsafety` on that will obviously cycle?

Reply:
No, that's not the problem. Let me illustrate with an example:
```rust
fn function() {
    let closure = || {};
}
```
We're trying to compute unsafeck for `closure`, which should invoke unsafeck for the parent body `function`. So `def.did` is the DefId of the closure, and the HIR is constructed in such a way that the corresponding HirId has `function` as its owner, so I though I could use that whenever I encounter a closure.

Enter consts:
```rust
fn function() -> [(); { || {}; 4 }] {}
```
Here, if I try the same "hack", i.e. get the HirId for the closure and take its owner, I get the DefId of `function` whereas I want the DefId of the anonymous constant `{ || {}; 4 }`.

To be fair I don't entirely understand which query invocations cycle and which don't when using `ensure`, and I've also not dived into the internals of HIR to check that `enclosing_body_owner` always does the right thing, but this PR doesn't cycle on the examples provided in #87414 and passed the rest of the test suite.

Reply:
Your explanation makes sense, thanks! I do think we should get a more solid strategy around all this, but I think it kind of overlaps with various other things like generics in anonymous constants. So let's merge this PR now